### DISCUSSION TITLE
Purpose of tensor_jacobian_product in adjoint optimization examples

### DISCUSSION THREAD

**Discussion by Andeloth** [NanoComp/meep#2098] [Q&A] [answered]
In some of the tutorials nlopt's gradient array is populated with simply the gradient returned by the adjoint solver, but in some a tensor_jacobian product is used, as shown below.

```
if gradient.size > 0:
        gradient[:] = tensor_jacobian_product(mapping,0)(v,eta_i,cur_beta,np.sum(dJ_du,axis=1)) # backprop
```

What is the purpose of doing that versus just populating it with the gradient?

Thanks!

**Comment by smartalecH** [comment, marked as answer]
> What is the purpose of doing that versus just populating it with the gradient?

Chain rule.

The adjoint solver doesn't know what other operations you perform on your design variables *before* you provide them to the solver itself. But your *optimization algorithm* needs to know the gradient w.r.t. the actual design parameters.

For example, if you filter and threshold your design variables ($\rho$), s.t. $\bar{\rho}=m(\rho)$ (where $m()$ is the final mapping function), the adjoint solver only ever sees $\bar{\rho}$. Which means the gradient it provides is $\frac{\partial f}{\partial \bar{\rho}}$, when you really want $\frac{\partial f}{\partial \rho}$.

So `tensor_jacobian_products` *backpropagates* your gradient (i.e. it just uses the chain rule).

[Our recent paper](https://doi.org/10.1364/OE.442074) discusses these nuances in detail (see section 4).
